sklearn/metrics/_classification.py

```diff
@@ -2949,16 +2920,22 @@ def log_loss(
             1 - transformed_labels, transformed_labels, axis=1
         )
 
-    # Clipping
-    y_pred = np.clip(y_pred, eps, 1 - eps)
-
     # If y_pred is of single dimension, assume y_true to be binary
     # and then check.
     if y_pred.ndim == 1:
         y_pred = y_pred[:, np.newaxis]
     if y_pred.shape[1] == 1:
         y_pred = np.append(1 - y_pred, y_pred, axis=1)
 
+    # Make sure y_pred is normalized
+    y_pred_sum = y_pred.sum(axis=1)
+    if not np.allclose(y_pred_sum, 1):
+        raise ValueError("The y_pred values do not sum to one.")
```

Comment:
Since we don't normalize internally, the result would completely meaningless. Better make sure to prevent bugs, no ?

Reply:
I also think that we should raise early if any y_pred < 0 or y_pred > 1, don't you ?

Reply:
I would prefer to start with a warning instead of an error:
- Probabilities not summing to 1 are not forbidden, just bad predictions (obviously uncalibrated).
- At what precision do we check against `1`. That's not so clear to me.

Reply:
Maybe we can try with `atol=10 * finfo(y_pred.dtype).eps`? That sounds reasonable to me. If people complains about a noisy warning, we can later decide to add a `warning_tol` parameter to control when to raise the warning (or even disable it entirely by setting it to `np.inf` or `None`).

Reply:
Alternatively, we can decide to not raise any or any error warning at all if the probabilities do not sum to one. The multiclass logloss value will not be minimal anyways for such models.

Reply:
@ogrisel +1

I would start with a warning because I think it‘s the best compromise for the time being. I would just set atol even a bit higher.

Reply:
I'm okay with a warning but I want to be clear that it's a change of mind, because we've already been having a warning for 2 versions now and the plan was to turn it into an error in 1.5:
```py
        warnings.warn(
            (
                "The y_pred values do not sum to one. Starting from 1.5 this"
                "will result in an error."
            ),
```

Reply:
I switched back to a warning, but if we don't intend to ever raise an error, maybe we should not warn either and just explain it in the docstring.

Reply:
Regarding the tolerance on the sum, I set `rtol=sqrt(eps)`. It's close to the tolerance we usually set in our tests (~1e-8 for float64, ~1e-4 for float32). I checked and none of our tests trigger the warning with this tol.

Reply:
Mmmh, I just don't want to raise an error. I actually like @jeremiedbb's current solution
> Regarding the tolerance on the sum, I set rtol=sqrt(eps). It's close to the tolerance we usually set in our tests (~1e-8 for float64, ~1e-4 for float32). I checked and none of our tests trigger the warning with this tol.

So still +1 for a warning. It gives us the freedom to remove it later or not. For a user, I consider it valuable information if proba does not some roughly to 1.

Reply:
Alright, then I think it's ready for a final review :)